File: experiments/bcut/bcut.rs
use std::io::{stdin, stdout, BufReader, BufWriter, Write, Read};
use std::env::args;

const MAX_COLUMNS: usize = 1 << 16;

#[inline]
fn read_i32(buf: &[u8]) -> usize {
    let mut out: usize = 0;
    let ptr_out = &mut out as *mut usize as *mut u8;
    unsafe {
        std::ptr::copy_nonoverlapping(buf.as_ptr(), ptr_out, 4);
    }
    out
}


#[inline]
fn read_u16(buf: &[u8]) -> usize {
    let mut out: usize = 0;
    let ptr_out = &mut out as *mut usize as *mut u8;
    unsafe {
        std::ptr::copy_nonoverlapping(buf.as_ptr(), ptr_out, 2);
    }
    out
}

fn main() {

    // parse args
    let fields: Vec<String> = args().collect();
    assert!(fields.len() == 2, "usage: bcut field1,field2,fieldN,...");
    let fields: Vec<usize> = fields[1]
        .split(",")
        .map(|x| x.parse::<i32>().unwrap())
        .map(|x| { assert!(x > 0 && x < MAX_COLUMNS as i32); (x - 1) as usize})
        .collect();

    // setup io
    let mut reader = BufReader::with_capacity(1024 * 512, stdin());
    let mut writer = BufWriter::with_capacity(1024 * 512, stdout());
    let mut buffer4: [u8; 4] = [0; 4];
    let mut chunk_offset: usize;
    let mut chunk_buffer: [u8; 1024*1024*5] = [0; 1024*1024*5];

    // setup state
    let mut max: usize;
    let mut sizes: [usize; MAX_COLUMNS] = [0; MAX_COLUMNS];
    let mut offsets: [usize; MAX_COLUMNS] = [0; MAX_COLUMNS];

    // process input line by line
    while let Ok(_) = reader.read_exact(&mut buffer4) {
        // read chunk size
        let chunk_size = read_i32(&buffer4);

        // read next chunk
        let mut chunk_buffer = &mut chunk_buffer[..chunk_size];
        reader.read_exact(&mut chunk_buffer).unwrap();

        // read all rows in chunk
        chunk_offset = 0;
        while chunk_offset < chunk_size {

            // read row max
            max = read_u16(&chunk_buffer[chunk_offset..]);
            chunk_offset += 2;

            // read row sizes
            for i in 0..max+1 {
                sizes[i] = read_u16(&chunk_buffer[chunk_offset..]);
                chunk_offset += 2;
            }

            // setup row offsets
            for i in 0..max+1 {
                offsets[i] = chunk_offset;
                chunk_offset += sizes[i] + 1;
            }

            // handle row
            let mut i = 0;
            for field in &fields {
                assert!(*field <= max, "found a row without enough columns");
                let offset = offsets[*field];
                let size = sizes[*field];
                writer.write_all(&chunk_buffer[offset..offset+size]).unwrap();
                i += 1;
                if i < fields.len() {
                    writer.write_all(&[b',']).unwrap();
                }
            }
            writer.write_all(&[b'\n']).unwrap();

        }
    }
}

File: experiments/bcut/bcut.go
package main

import (
	"bufio"
	"encoding/binary"
	"io"
	"os"
	"strconv"
	"strings"
)

func main() {
	var fields []int
	for _, x := range strings.Split(os.Args[1], ",") {
		x, err := strconv.Atoi(x)
		if err != nil {
			panic(err)
		}
		fields = append(fields, x-1)
	}
	max := int32(0)
	sizes := make([]int32, 1<<16)
	offsets := make([]int32, 1<<16)
	r := bufio.NewReader(os.Stdin)
	w := bufio.NewWriter(os.Stdout)
	defer w.Flush()
	// buffer4 := make([]byte, 4)
	chunk_offset := int32(0)
	chunk_size := int32(0)
	chunk_buffer := make([]byte, 1024*1024*5)
	for {

		// read chunk size
		err := binary.Read(r, binary.LittleEndian, &chunk_size)
		if err != nil {
			break
		}

		// read chunk
		_, err = io.ReadFull(r, chunk_buffer[:chunk_size])
		if err != nil {
			panic(err)
		}

		// read all rows in chunk
		chunk_offset = 0
		for chunk_offset < chunk_size {
			// read row max
			max = int32(binary.LittleEndian.Uint16(chunk_buffer[chunk_offset:]))
			chunk_offset += 2

			// read row sizes
			for i := int32(0); i <= max; i++ {
				sizes[i] = int32(binary.LittleEndian.Uint16(chunk_buffer[chunk_offset:]))
				chunk_offset += 2
			}

			// setup row offsets
			for i := int32(0); i <= max; i++ {
				offsets[i] = chunk_offset
				chunk_offset += sizes[i] + 1
			}

			// handle row
			for i, f := range fields {
				_, err = w.Write(chunk_buffer[offsets[f] : offsets[f]+sizes[f]])
				if err != nil {
					panic(err)
				}
				if i != len(fields)-1 {
					_, err = w.Write([]byte(","))
					if err != nil {
						panic(err)
					}
				}
			}
			_, err = w.Write([]byte("\n"))
			if err != nil {
				panic(err)
			}
		}
	}

}
